doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +668,381 @@ Applying theorems
+
+   .. insertprodn in_hyp_as as_ipat
+
+   .. prodn::
+      in_hyp_as ::= in {+, @ident {? @as_ipat } }
+      as_ipat ::= as @simple_intropattern
+
+   Uses unification to match the type of each :n:`@one_term`
+   (in :n:`@one_term_with_bindings`) with the goal
+   (to do :term:`backward reasoning`) or with a hypothesis (to do :term:`forward reasoning`).
+   Specifying multiple :n:`@one_term_with_bindings` is equivalent to
+   giving each one serially, left to right, as separate `apply` tactics.
+
+   The type of :n:`@one_term` contains zero or more :term:`premises <premise>`
+   followed by a :ref:`conclusion <conclusion_meaning_2>`,
+   i.e. it typically has the form :n:`{? forall @open_binders , } {* @term__premise -> } @term__conclusion`.
+   (The ``forall``\s may also be interleaved with the premises, but common usage is
+   to equivalently gather them at the beginning of the :n:`@one_term`.)
+   Backward reasoning with a :n:`@one_term` whose type is, for example, `A -> B`
+   replaces an as-yet unproven goal `B` with `A`.  Forward reasoning with the same
+   :n:`@one_term` changes a hypothesis with type `A` to `B`.  (Hypotheses are
+   considered proven propositions within the context that contains them.)
+
+   Unification creates a map from the variables in the type of :n:`@one_term`
+   to matching subterms of the goal or hypothesis.
+   The matching subterms are then substituted into the type of :n:`@one_term`
+   when generating the updated goal or hypothesis.  Unmatched premises become
+   new subgoals with similar substitutions.  If no match is found, the
+   tactic fails.
+
+   The goal and hypothesis cases are described separately for clarity.
+
+.. _unused1:
+
+   .. the dummy ref name is needed to get correct formatting of the next line and "Without..."
+
+   :n:`@one_term` (inside :n:`@one_term_with_bindings`)
+     If :n:`@one_term` is an :n:`@ident`, it is the name of
+     a theorem, lemma or hypothesis whose type is given in the
+     theorem statement or shown in the context.  Otherwise it is a proof term whose
+     type can be displayed with :cmd:`Check` :n:`@one_term`.
+
+   Without :n:`@in_hyp_as` (the goal case)
+     If the goal matches all of the type of :n:`@one_term` (both premises and
+     the conclusion), the tactic proves the goal.
+     Otherwise, the tactic matches the goal against the conclusion of :n:`@one_term`
+     and, if possible, one or more premises (from right to left).
+     If the match succeeds, the tactic replaces the current goal with a subgoal for
+     each unmatched premise of the type of :n:`@one_term`.  This
+     :ref:`example <apply_backward>` matches only the conclusion, while
+     this :ref:`one <apply_backward_w_premises>` also matches a premise.
+
+     If the conclusion of the type of :token:`one_term` does not match the goal
+     *and* the conclusion is an inductive type with a single constructor,
+     then each premise in the constructor is recursively matched to the goal in
```

Comment:
The constructor has a type which has premises, but the constructor itself does not have premises, it has arguments (when it is applied). So maybe this should be:
```suggestion
     then each premise in the constructor's type is recursively matched to the goal in
```

Though, even after this change, I'm still a bit uncertain that this is the right wording, because the premises of the constructor's type are somewhat abstract and independent of the context in which the inductive type was used. But I don't have any suggestion to improve this further, and I'm otherwise relatively convinced that your removal of the undefined "isomorphic to a tuple type" concept is a good idea.

Reply:
How about "argument" instead of "premise"?  IIUC `Inductive` shows the types, so should be quite clear.

Reply:
Yes, I pondered whether to recommend using "argument" instead of "premise". However, it's not an argument of the constructor but an argument of the inductive type itself that is recursively matched to the goal (the constructor will be eventually found in the proof term, applied to subgoals' proof terms, not to subgoals themselves).

I also was a bit worried by the case of the `eq` type, which is an inductive type with a single constructor, but not isomorphic to a tuple type (its constructor does not take any argument), but it's probably OK to ignore this edge case...

Reply:
No, it is the type of the arguments of the sole constructor.

Reply:
Thanks. But then "type of the arguments of the sole constructor" is pretty much what "premise in the constructor's type" meant, or is there any difference?

Reply:
I guess it is the same.

Reply:
So @jfehrle, let's leave things as they are (unless someone comes up with a simpler wording, which would be greatly appreciated).